Created new Vegetation Gem Reference doc (#1923)
o3de/o3de.org: content/docs/user-guide/gems/reference/environment/vegetation.md

---
linkTitle: Vegetation
title: Vegetation Gem
description: The Vegetation Gem provides tools to place natural-looking vegetation in Open 3D Engine.
weight: 400
toc: true
---

> The **Vegetation Gem** provides tools for procedurally populating landscapes and environments. These tools comprise of system and editor components that automate the placement of static objects using a data-driven approach. You can use these tools to replace or compliment manually painting, editing, and saving every instance in your level. 

Reviewer: "painting" -> "placing". Lumberyard had a static vegetation painting workflow, but O3DE does not, so the current alternative to using this Gem would be to manually place every instance in the level yourself.

Reviewer: "automate the placement of static objects" - this might be a bit misleading. The system dynamically spawns and despawns the objects around the camera as the camera moves around, so I'm worried that "placement of static objects" might give the wrong impression that this is an edit-time-only tool that generates instances that are statically saved with the level. Maybe "automate the selection, placement and management of vegetation objects dynamically at runtime"?

Author: Rephrased to: 

```suggestion
The **Vegetation Gem** provides tools for procedurally populating landscapes and environments. These tools comprise of system and editor components that use a data-driven approach to automatically select, place, and manage vegetation objects dynamically at runtime. You can use these tools to replace or compliment manually placing, editing, and saving every instance in your level. 
```

Many features of the dynamic vegetation system rely on other Gems and components to supply data about the environment, such as surfaces, images, and gradient signals, to direct where and how vegetation appears. 

## Dependencies

- [Surface Data Gem]()
- [Gradient Signal Gem]()
- [FastNoise Gem]() (optional)

> The Surface Data Gem allows a surface, like terrain or meshes, to emit signals or tags that communicate its surface type. For example, with a Vegetation Surface Mask Filter component, you can choose which vegetation to grow on a particular surface by using an inclusion and exclusion list. You can also recapture the tags as a gradient signal by using the Surface Mask Gradient component.

Reviewer: "which vegetation to grow" -> "which types of vegetation can be placed"

The Gradient Signal Gem provides components that direct data to the vegetation system, controlling the appearance of dynamic vegetation. Using gradient signals with vegetation modifiers or filters, such as Position Modifier and Distribution Filter components, produce realistic, random expressions of vegetation in the world.

## Provides

Source code: [`/Gems/docs/user-guide/components/reference/Vegetation/`](https://github.com/o3de/o3de/tree/development/Gems//docs/user-guide/components/reference/Vegetation/Code)

API: [Vegetation Gem API Reference](https://www.o3de.org/docs/api/gems//docs/user-guide/components/reference/vegetation/index.html)

System components:
- [Vegetation Area System](/docs/user-guide/components/reference/docs/user-guide/components/reference/vegetation/vegetation-area-system-component)
- [Vegetation Instance System](/docs/user-guide/components/reference/docs/user-guide/components/reference/vegetation/vegetation-instance-system-component)

> 

Reviewer: We should also add in a section for the Level components - Vegetation System Settings and Vegetation Debugger.

> Editor components:

Reviewer: Maybe "Entity components"? They're not just Editor components because they also exist at runtime.

Other systems in the engine register and process Vegetation areas.
They can examine and take action on points in the world, like placing or preventing a vegetation instance.

In O3DE Editor, vegetation areas are referred to as *vegetation layers*.
You can add vegetation behaviors to vegetation layers using components that describe, select, reject, or mutate potential vegetation instances. 
These are called vegetation descriptor providers, selectors, filters, and modifiers respectively. 
Each has an interface that extends the system with new features. 
The Vegetation Gem contains several versions of each type that each has a unique purpose and effect on generated vegetation instances. 
These components expose controls that allow rapid, rule-based, procedural population of any arbitrarily sized part of the world with decorative content. 
Depending on its configuration, a single Vegetation Layer can act as a small patch of flowers or intelligently blanket the entire world with a convincing variety and groupings of plants and objects.

## Vegetation instances

> *Vegetation instances* are the objects placed throughout the world. They're created using vegetation descriptors. 

Author: ```suggestion
*Vegetation instances* are the static objects placed throughout the world. They're created using vegetation descriptors. 
```

Reviewer: "Static" is misleading. They're dynamically created and destroyed at runtime, and they can even technically be moving objects. For example, if you put a sphere rigid body into a prefab, you can use the vegetation system to spawn them and they will start rolling around as they appear.

They contain a unique ID, transform, other attributes, and a reference to the source vegetation descriptor.

In O3DE Editor, vegetation instances appear within a vegetation area. They-re procedurally generated based on the vegetation behaviors that you can add to the vegetation area. 

> ## Descriptors

Author: ```suggestion
## Vegetation descriptors
```

> *Vegetation descriptors* are structures that specify all of the common details needed to represent a type of vegetation. It includes data for mesh and material assets, the type of vegetation render node that should be created, the affects of wind and physics, and many advanced parameters that can be enabled to override behavior of most of the included filters and modifiers. **Vegetation Descriptors** can be created in the **Vegetation Asset List** component or through the Asset Editor by creating and editing **Vegetation Descriptor List Assets**.

Reviewer: "the affects of" -> "the effects of".

Author: Should this paragraph also be changed to remove mentions of "render nodes" and "wind and physics"? Since they are deprecated?

Reviewer: Yep, absolutely, great catch! "render node" -> "instance", and remove "the effects of wind and physics", since those aren't specified here any more.

|     | **Surface Slope Alignment Min** | Override minimum alignment strength. | Float |
|     | **Surface Slope Alignment Max** | Override maximum alignment strength. | Float |
| Slope Filter |     |     |     |
|     | **Slope Filter Override Enabled** | When enabled, the slope filter component can use the per instance override values specified in the descriptor instead of the component configuration. | Bool |
|     | **Slope Filter Min** | Override minimum slope accepted by filter. | Float |
|     | **Slope Filter Max** | Override maximum slope accepted by filter. | Float |
| Surface Mask Filter |     |     |     |
|     | **Override Mode** | Controls how overrides are considered by the surface mask filter component<br><br>Disable - Overrides are completely ignored<br><br>Replace - Overrides replace those specified in the component<br><br>Extend - Overrides are added to the specified in the component. |     |
|     | **Inclusion Tags** | A set of surface tags that dictate where vegetation can be placed. | SurfaceTagVector |

> |     | **Exclusion Tags** | A set of surface tags that dictate where vegetation can not be placed. | SurfaceTagVector |

Reviewer: "can not" -> might sound better as "will not"

| Surface Mask Depth Filter |     |     |     |
|     | **Surface Tags** | A set of surface tags used for depth comparisons. | SurfaceTagVector |
|     | **Upper Distance Range (m)** | A range used for vertical distance comparisons against the closest point with matching surface tags. | Float |
|     | **Lower Distance Range (m)** | A range used for vertical distance comparisons against the closest point with matching surface tags. | Float |

File: [`/Gems/Vegetation/Code/Source/Descriptor.cpp`](https://github.com/o3de/o3de/blob/900bd390b9b11c83a77ac3de651dc9eb3d897270/Gems/Vegetation/Code/Include/Vegetation/Descriptor.h)

## EBus interface

### Vegetation::AreaInfoBus

| Request Name | Description | Parameters | Return |
| --- | --- | --- | --- |

> | `GetLayer` |     | None | Float |

Author: Comment in code: "to handle overlapping vegetation layers"
@mbalfour-amzn Can you further explain the use of this function?

Reviewer: "Layer" and "Priority" are two different priority fields. "Layer" is a macro priority (foreground / background), and "Priority" is a micro priority within a layer. Some examples:
* Foreground / 0 is higher priority than Background / 0.
* Foreground / 10 is higher priority than Foreground / 0.
* Foreground / 0 is higher priority than Background / 10.

The purpose of the priorities is that it's possible to create multiple vegetation areas that overlap in the world. When that occurs, for the overlapped area, the highest-priority vegetation area is given the opportunity to spawn (or block) vegetation instances. The other vegetation areas are then given turns in decreasing priority order until there's no more room to spawn anything, or no more vegetation areas to process.

GetLayer() returns the value of the macro priority (the "layer")
GetPriority returns the value of the micro priority (the "priority")

> | `GetPriority` | Gets the combined layer and area priority value as a single combined global priority. | None | Float |

Reviewer: It doesn't actually return a combined global priority. It only returns the micro priority value.

| `GetEncompassingAabb` | Gets the axis-aligned bounding box for the entire vegetation area. | None | AZ::Aabb |

> | `GetProductCount` | Gets the number of currently spawned instances in the vegetation area. | None | AZ::u32 |

Reviewer: I _think_ this is actually "Gets the number of instances currently spawned by this vegetation area." The subtle distinction is that if you have two areas that overlap, you could have instances from that second area that are physically _in_ the first area but weren't _spawned_ by that first area.

| `GetChangeIndex` | Gets an incrementing number that represents the number of times the Blocker area refreshed since creation. | None | AZ::u32 |

File: [`/Gems/Vegetation/Code/Include/Vegetation/Ebuses/AreaInfoBus.h`](https://github.com/o3de/o3de/blob/0121a0675b53104cfe1d9f6d87bf9b8ac59a4382/Gems/Vegetation/Code/Include/Vegetation/Ebuses/AreaInfoBus.h)

### Vegetation::AreaRequestBus

| Request Name | Description | Parameters | Return |
| --- | --- | --- | --- |

> | `PrepareToClaim` | Runs any pre-claim checks or logic that is not needed per point. | EntityIdStack& stackIds | Bool |

Reviewer: "Runs any pre-claim checks or logic that is not needed per point" -> This might make more sense as "Runs any position-independent pre-claim checks or logic"

|     | **Altitude Filter Override Enabled** | When enabled, the altitude filter component can use the per instance override values specified in the descriptor instead of the component configuration. | Bool |
|     | **Altitude Filter Min** | Override value for the minimum altitude accepted by the filter. | Float |
|     | **Altitude Filter Max** | Override value for the maximum altitude accepted by the filter. | Float |
| Distance Between Filter (Radius) |     | Settings for controlling per instance overrides for the distance between filter component. |     |
|     | **Distance Between Filter Override Enabled** | When enabled, the distance between filter component can use the per instance override values specified in the descriptor instead of the component configuration. | Bool |
|     | **Bound Mode** | When performing distance checks between two instances, this setting determines whether the radius of the mesh asset is used vs a manually entered radius<br><br>Radius<br><br>MeshRadius. |     |
|     | **Radius Min** | User defined radius for distance checks. | Float |

> | Surface Slope Alignment |     |     |     |

Reviewer 2: ```suggestion
| Surface Slope Alignment |     |     |     |
```
This and the rest of the component overrides don't have a property group description.  I don't think in any case it is a major issue, we could unify their treatment it at some point (or not). I have left property group descriptions blank like this in many other components.

Author: I actually want to reformat this so it matches the standard "properties" formatting. With that format, the group names will be formatted as headers, instead of having a column like this.

|     | **Surface Slope Alignment Override Enabled** | When enabled, the slope alignment modifier component can use the per instance override values specified in the descriptor instead of the component configuration. | Bool |
|     | **Surface Slope Alignment Min** | Override minimum alignment strength. | Float |
|     | **Surface Slope Alignment Max** | Override maximum alignment strength. | Float |
| Slope Filter |     |     |     |

> |     | **Slope Filter Override Enabled** | When enabled, the slope filter component can use the per instance override values specified in the descriptor instead of the component configuration. | Bool |

Reviewer 2: ```suggestion
|     | **Slope Filter Override Enabled** | When enabled, the [Slope Filter](/docs/user-guide/components/reference/vegetation-filters/vegetation-slope-filter) component can use the per-instance override values specified in the descriptor instead of the component configuration. | Bool |
```

|     | **Slope Filter Min** | Override minimum slope accepted by filter. | Float |
|     | **Slope Filter Max** | Override maximum slope accepted by filter. | Float |
| Surface Mask Filter |     |     |     |

> |     | **Override Mode** | Controls how overrides are considered by the surface mask filter component<br><br>Disable - Overrides are completely ignored<br><br>Replace - Overrides replace those specified in the component<br><br>Extend - Overrides are added to the specified in the component. |     |

Reviewer 2: ```suggestion
|     | **Override Mode** | Controls how overrides are considered by the [Surface Mask Filter](/docs/user-guide/components/reference/vegetation-filters/vegetation-surface-mask-filter) component<br><br>Disable - Overrides are completely ignored<br><br>Replace - Overrides replace those specified in the component<br><br>Extend - Overrides are added to the specified in the component. |     |
```

|     | **Inclusion Tags** | A set of surface tags that dictate where vegetation can be placed. | SurfaceTagVector |
|     | **Exclusion Tags** | A set of surface tags that dictate where vegetation will not be placed. | SurfaceTagVector |

> | Surface Mask Depth Filter |     |     |     |

Reviewer 2: ```suggestion
| Surface Mask Depth Filter |   |  Settings for controlling per-instance overrides for the [Surface Mask Depth Filter](/docs/user-guide/components/reference/vegetation-filters/vegetation-surface-mask-depth-filter) component.  |     |
```
Added this group property description because the individual properties lacked a link to the relevant component.

| `GetChangeIndex` | Gets an incrementing number that represents the number of times the Blocker area refreshed since creation. | None | AZ::u32 |

File: [`/Gems/Vegetation/Code/Include/Vegetation/Ebuses/AreaInfoBus.h`](https://github.com/o3de/o3de/blob/0121a0675b53104cfe1d9f6d87bf9b8ac59a4382/Gems/Vegetation/Code/Include/Vegetation/Ebuses/AreaInfoBus.h)

### Vegetation::AreaRequestBus

| Request Name | Description | Parameters | Return |
| --- | --- | --- | --- |
| `PrepareToClaim` | Runs any pre-claim checks or logic, independent of position. | EntityIdStack& stackIds | Bool |
| `ClaimPositions` | Processes a set of points for planting vegetation or performing other operations. | EntityIdStack& stackIds, ClaimContext& context | None |
| `UnclaimPosition` | Handles clean up whenever the system releases a claimed point. | const ClaimHandle handle | None |

File: [`/Gems/Vegetation/Code/Include/Vegetation/Ebuses/AreaRequestBus.h`](https://github.com/o3de/o3de/blob/0fb9727c67c9ad7885b4af538860920f5ba53bfa/Gems/Vegetation/Code/Include/Vegetation/Ebuses/AreaRequestBus.h)

### Vegetation::DescriptorProviderRequestBus

> *Vegetation descriptor providers* supply Vegetation Layer Spawners and Blenders with a list of vegetation descriptors from defined sources. 

Reviewer 2: ```suggestion
*Vegetation descriptor providers* supply [Vegetation Layer Spawners](/docs/user-guide/components/reference/docs/user-guide/components/reference/vegetation/layer-spawner) and [Blenders](/docs/user-guide/components/reference/docs/user-guide/components/reference/vegetation/vegetation-layer-blender) with a list of vegetation descriptors from defined sources. 
```

The Vegetation Gem comes with components that can either provide descriptors from a list that's defined directly in the component, reference a list of descriptors from externally created assets, or combine multiple descriptor lists together.

| Request Name | Description | Parameters | Return |
| --- | --- | --- | ---  |
| `GetDescriptors` | When a component is ready and has fully loaded all of its assets, this method creates a list of its active descriptors. | DescriptorPtrVec& descriptors | None |

File: [`/Gems/Vegetation/Code/Include/Vegetation/Ebuses/DescriptorProviderRequestBus.h`](https://github.com/o3de/o3de/blob/0fb9727c67c9ad7885b4af538860920f5ba53bfa/Gems/Vegetation/Code/Include/Vegetation/Ebuses/DescriptorProviderRequestBus.h)

### Vegetation::DescriptorSelectorRequestBus

> A Vegetation Layer Spawner component can use *vegetation descriptor selectors* when a vegetation descriptor Provider offers multiple options. There is currently one selector, Vegetation Asset Weight Selector component, which chooses based on the selection weight field of a vegetation descriptor.

Reviewer 2: ```suggestion
A [Vegetation Layer Spawner](/docs/user-guide/components/reference/docs/user-guide/components/reference/vegetation/layer-spawner) component can use *vegetation descriptor selectors* when a vegetation descriptor provider offers multiple options. There is one selector, [Vegetation Asset Weight Selector](/docs/user-guide/components/reference/docs/user-guide/components/reference/vegetation/vegetation-asset-weight-selector)component, which chooses based on the selection weight field of a vegetation descriptor.
```

| Request Name | Description | Parameters | Return|
| --- | --- | --- | ---  |

> | `SelectDescriptors` | Uses the input gradient signal and other parameters to reduce the set of descriptors to those eligible for selection | None | const DescriptorSelectorParams&  <br>DescriptorPtrVec&|

Reviewer 2: ```suggestion
| `SelectDescriptors` | Uses the input gradient signal and other parameters to reduce the set of descriptors to those eligible for selection. | None | const DescriptorSelectorParams&  <br>DescriptorPtrVec&|
```

File: [`/Gems/Vegetation/Code/Include/Vegetation/Ebuses/DescriptorSelectorRequestBus.h`](https://github.com/o3de/o3de/blob/0fb9727c67c9ad7885b4af538860920f5ba53bfa/Gems/Vegetation/Code/Include/Vegetation/Ebuses/DescriptorSelectorRequestBus.h)

### Vegetation::ModifierRequestBus

*Vegetation modifiers* add uniqueness or variation to every vegetation instance, by changing the position, rotation, scale, alignment, or any exposed field. 
This allows the same vegetation descriptors throughout a vegetation area, with each instance of the descriptor to appear differently.

| Request Name | Description | Parameters | Return |
| --- | --- | --- | ---   |

> | `Execute` | Modify a single vegetation instance based on configuration | Vegetation::InstanceData | None |

Reviewer 2: ```suggestion
| `Execute` | Modify a single vegetation instance based on configuration. | Vegetation::InstanceData | None |
```

| `GetModifierStage` | Internal: Determines the order in which to run Vegetation Modifier components. | None | Vegetation::ModifierStage |

File: [`/Gems/Vegetation/Code/Include/Vegetation/Ebuses/ModifierRequestBus.h`](https://github.com/o3de/o3de/blob/0fb9727c67c9ad7885b4af538860920f5ba53bfa/Gems/Vegetation/Code/Include/Vegetation/Ebuses/ModifierRequestBus.h)

### Vegetation::FilterRequestBus

Vegetation areas use *vegetation filters* to evaluate every vegetation instance data and decide whether any activity should occur. 

> Vegetation Layer Spawners use filters to determine whether to create vegetation instances. 

Reviewer 2: ```suggestion
[Vegetation Layer Spawners](/docs/user-guide/components/reference/docs/user-guide/components/reference/vegetation/layer-spawner) use filters to determine whether to create vegetation instances. 
```
